File: ios/positivelys/AppService.swift
//
// Created by Logan Keenan on 11/24/20.
//

import Foundation

public class AppService {

    public static var hostName: String = "https://positivelys.com";

    func appContext() -> String {
        let database_path: String = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0] + "/database.sqlite"
        let local_files_path: String = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0];
        let assets_path = Bundle.main.bundlePath
        let views_path = "\(Bundle.main.bundlePath)/dist/views/"

        #if DEBUG
        let environment = "development"
        #else
        let environment = "production"
        #endif

        let appContext = AppContext(
                database_path: database_path,
                local_files_path: local_files_path,
                assets_path: assets_path,
                views_path: views_path,
                environment: environment,
                os: "ios"
        )

        return appContext.asJson()
    }

    public func make_request(appRequest: AppRequest) -> AppResponse {
        let appRequestData = try! JSONEncoder().encode(appRequest)
        let appRequestAsJson = String(data: appRequestData, encoding: .utf8)

        let result = make_app_request(appRequestAsJson, appContext())
        let app_response_as_json = String(cString: result!)

        make_app_request_free(UnsafeMutablePointer(mutating: result))

        let response: AppResponse? = try? JSONDecoder().decode(AppResponse.self, from: app_response_as_json.data(using: .utf8)!)

        return response!;
    }

}

File: android/app/src/main/java/com/cultivatedsoftware/positivelys/services/AppService.kt
package com.cultivatedsoftware.positivelys.services

import com.cultivatedsoftware.positivelys.models.AppContext
import com.cultivatedsoftware.positivelys.models.AppRequest
import com.cultivatedsoftware.positivelys.models.AppResponse
import com.google.gson.Gson

class AppService(var dataDirectory: String, var filesDirectory: String) {

    var gson: Gson = Gson();
    fun makeRequest(appRequest: AppRequest): AppResponse {
        val appRequestAsJson = gson.toJson(appRequest)
        val appContext = AppContext(
            database_path = "$dataDirectory/database.sqlite",
            assets_path = "file:///android_asset",
            local_files_path = "$filesDirectory",
            views_path = "$dataDirectory/dist/views/",
            environment = "development",
            os = "android"
        )
        val appContextAsJson = gson.toJson(appContext)

        val appResponseAsJson = makeapprequest(appRequestAsJson, appContextAsJson)
        val response = gson.fromJson<AppResponse>(appResponseAsJson, AppResponse::class.java)

        return response
    }

    init {
        System.loadLibrary("rust_android")
    }

    external fun makeapprequest(j_app_request: String, j_app_context: String): String
}